aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+    // we do this post-processing only if 'vectors' is complex-valued
+    // otherwise storage of 'vectors' is used directly
+    if (vectors.is_complex() && compute_eigenvectors) {
+      maybe_complex_vectors = at::empty(input.sizes(), options, MemoryFormat::Contiguous);
+      maybe_complex_vectors.transpose_(-2, -1);  // make 'maybe_complex_vectors' to have Fortran contiguous memory layout
+    }
+  }
+
+  // now call the actual computation
+  linalg_eig_stub(input.device().type(), real_imag_values, maybe_complex_vectors, infos, input, compute_eigenvectors);
+
+  // if input is not complex we need to do some post-processing
+  if (!input.is_complex()) {
+    // extract real and imaginary parts of the output
+    auto real_values = real_imag_values.slice(/*dim=*/-1, /*start=*/0, /*end*/input.size(-1));
+    auto imag_values = real_imag_values.slice(/*dim=*/-1, /*start=*/input.size(-1));
+
+    // if the imaginary part is zero we don't need to do anything
+    bool is_zero_imag = at::all(imag_values == 0.0).item().toBool();
+    if (is_zero_imag) {
+      values.copy_(real_values);
+      if (compute_eigenvectors) {
+        vectors.copy_(maybe_complex_vectors);  // does nothing for !vectors.is_complex() because vectors.is_same(maybe_complex_vectors) == true
+      }
+      return std::tuple<Tensor&, Tensor&>(values, vectors);
+    }
+
+    if (values.is_complex()) {
+      values = at::complex_out(values, real_values, imag_values);
+    } else {
+      TORCH_CHECK(false, "linalg_eig: imaginary part of eigenvalues is non-zero, can't safely cast eigenvalues to non-complex dtype.")
+    }
+    if (compute_eigenvectors) {
+      if (vectors.is_complex()) {
+          vectors = linalg_eig_make_complex_eigenvectors(vectors, values, maybe_complex_vectors);
+      } else {
+        TORCH_CHECK(false, "linalg_eig: imaginary part of eigenvectors is non-zero, can't safely cast eigenvectors to non-complex dtype.")
+      }
+    }
+  }
+
+  return std::tuple<Tensor&, Tensor&>(values, vectors);
+}
+
+std::tuple<Tensor&, Tensor&> linalg_eig_out(const Tensor& input, Tensor& values, Tensor& vectors) {
+  squareCheckInputs(input);
+
+  checkLinalgCompatibleDtype("linalg_eig", values, input, "eigenvalues");
+  checkLinalgCompatibleDtype("linalg_eig", vectors, input, "eigenvectors");
+
+  // MAGMA doesn't have GPU interface for GEEV routine, it requires inputs to be on CPU
+  auto options = input.options().device(at::kCPU);
+  auto infos = at::zeros({std::max<int64_t>(1, batchCount(input))}, options.dtype(kInt));
+
+  // if result is not empty and not in batched column major format we have to allocate a temporary tensor
+  bool is_batched_column_major = false;
+  if (vectors.dim() >= 2) {
+    is_batched_column_major = vectors.transpose(-2, -1).is_contiguous();
+  }
+
+  // for NumPy compatibility we can't determine the dtype of eigenvalues before running the computation
```

Comment:
Let's talk more about this.

So in NumPy if the input is real-valued then the dtype of the returned tensor is real if the eigenvalues are real-valued and complex if the eigenvalues have non-zero imaginary parts?

How is that actually implemented in NumPy?

Should we mimic this behavior? PyTorch tends to prefer static shape propagation. That is, we won't be able to write shape propagation logic for this operation if its output dtype is value-dependent. Note that JAX does NOT implement this NumPy behavior either (they're very dependent on static shape propagation):

```
np.linalg.eig(np.diag((1, 2, 3)))
: (array([1., 2., 3.]), array([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]]))

jnp.linalg.eig(np.diag((1, 2, 3)))
: [DeviceArray([1.+0.j, 2.+0.j, 3.+0.j], dtype=complex64),
 DeviceArray([[1.+0.j, 0.+0.j, 0.+0.j],
              [0.+0.j, 1.+0.j, 0.+0.j],
              [0.+0.j, 0.+0.j, 1.+0.j]], dtype=complex64)]
```

cc @rgommers, this is another interesting example of shape propagation logic considerations that the array-api probably wants to consider

Reply:
> So in NumPy if the input is real-valued then the dtype of the returned tensor is real if the eigenvalues are real-valued and complex if the eigenvalues have non-zero imaginary parts?
How is that actually implemented in NumPy?

That's right. NumPy checks whether all values of the imaginary part of eigenvalues are equal to 0.0 with no tolerance, if that's the case casts output to reals.
https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/linalg/linalg.py#L1325

In Julia, it's the same behavior (all the imaginary values of eigenvalues are also compared to be == 0)
https://github.com/JuliaLang/julia/blob/788b2c77c10c2160f4794a4d4b6b81a95a90940c/stdlib/LinearAlgebra/src/eigen.jl#L157

Reply:
PyTorch's correct behavior would be easier to determine if we were further along in our program acquisition work. For now I would suggest we follow JAX and warn that this behavior may change in the future, however. That is, I suspect our method of program acquisition will require shape propagation logic to be value-independent.